File: frontend/src/mocks/otTeam.js
/* Demo-mode OT team: team roles, the directory of outside doctors / partners, and the checks on a
   case's team (billing.team). Same method names and shapes as `otTeam` in src/api/real.js; the
   server rules are in backend/app/services/ot_team.py.
   State lives in `store.state.otTeamLists`, created on first use and re-created after
   `store.reset()` (which swaps the otCases array the state is tied to). */
import { store, latency } from './store';
import { printSettings } from './rxPrint';

const S = store.state;
const c = store.clone;

// Same starting roles as the server seed (backend/app/seed/ot_team.py); fees start at ₹0.
const SEED_ROLES = [
  ['surgeon', 'Surgeon'],
  ['assistant_surgeon', 'Assistant surgeon'],
  ['anaesthetist', 'Anaesthetist'],
  ['scrub_nurse', 'Scrub nurse'],
  ['ot_technician', 'OT technician'],
];
const STAFF_ROLES_IN_OT = ['doctor', 'ot_staff', 'optometrist'];

function state() {
  if (!S.otTeamLists || S.otTeamLists.owner !== S.otCases) {
    S.otTeamLists = {
      owner: S.otCases,
      roles: SEED_ROLES.map(([key, label], i) => ({ id: i + 1, key, label, defaultFee: 0, sortOrder: i, active: true })),
      partners: [
        {
          id: 1,
          name: 'Dr. Kavita Shah',
          qualification: 'MD Anaesthesia',
          regNo: 'G-24518',
          phone: '9824000011',
          defaultRoleKey: 'anaesthetist',
          defaultFee: 2500,
          note: 'Visiting anaesthetist',
          active: true,
          createdAt: new Date().toISOString(),
        },
      ],
    };
  }
  return S.otTeamLists;
}

function httpError(status, message) {
  const err = new Error(message);
  err.response = { status, data: { detail: message } };
  return err;
}

const text = (v) =>
  String(v ?? '')
    .replace(/\s+/g, ' ')
    .trim();
const ordered = (rows) => rows.slice().sort((a, b) => a.sortOrder - b.sortOrder || a.id - b.id);
const byName = (rows) => rows.slice().sort((a, b) => a.name.toLowerCase().localeCompare(b.name.toLowerCase()));

function fee(v, where) {
  if (v == null || v === '') return 0;
  let n = v;
  if (typeof v === 'string') {
    const s = v.replace(/[,₹\s]/g, '');
    if (!/^-?\d+$/.test(s)) throw httpError(422, `${where}: the fee must be whole rupees, e.g. 2500`);
    n = Number(s);
  }
  if (typeof n !== 'number' || Number.isNaN(n)) throw httpError(422, `${where}: the fee must be a number of rupees`);
  if (!Number.isInteger(n)) throw httpError(422, `${where}: the fee must be whole rupees (no paise)`);
  if (n < 0) throw httpError(422, `${where}: the fee cannot be below ₹0`);
  return n;
}

/* ---------------- the team on a case (used by the demo OT adapter) ---------------- */
const rowEmpty = (r) => !['name', 'qualification', 'regNo'].some((k) => text(r?.[k])) && !Number(r?.fee);

/** Validate + tidy a team like the server: fully empty rows dropped; known role, a name, whole-rupee
    fee ≥ 0; an outside member needs a qualification. Throws a 422 naming the row. */
export function cleanTeam(team) {
  if (team == null) return [];
  if (!Array.isArray(team)) throw httpError(422, 'The OT team must be a list of team members');
  const roles = state().roles;
  const out = [];
  team.forEach((raw, idx) => {
    const i = idx + 1;
    if (rowEmpty(raw)) return;
    const key = text(raw.roleKey);
    const role = roles.find((r) => r.key === key);
    let where = `OT team row ${i}${role ? ` (${role.label})` : ''}`;
    if (!key) throw httpError(422, `${where}: pick a role (Surgeon, Anaesthetist…)`);
    if (!role) throw httpError(422, `${where}: unknown role '${key}'`);
    const name = text(raw.name);
    if (!name) throw httpError(422, `${where}: enter the person's name`);
    where = `OT team row ${i} (${role.label}, ${name})`;
    const qualification = text(raw.qualification);
    const external = !!raw.external;
    if (external && !qualification)
      throw httpError(422, `${where}: an outside doctor needs a medical qualification (e.g. MD Anaesthesia)`);
    const partnerId = raw.partnerId == null || raw.partnerId === '' ? null : Number(raw.partnerId);
    if (partnerId != null && !state().partners.some((p) => p.id === partnerId))
      throw httpError(422, `${where}: unknown outside doctor`);
    out.push({
      roleKey: key,
      roleLabel: text(raw.roleLabel) || role.label,
      name,
      qualification,
      regNo: text(raw.regNo),
      external,
      partnerId,
      fee: fee(raw.fee, where),
    });
  });
  return out;
}

/** A new surgery starts with the clinic's doctor as Surgeon (from the prescription print settings). */
export function defaultTeam() {
  const role = state().roles.find((r) => r.key === 'surgeon');
  const s = printSettings();
  if (!role || !role.active || !text(s.doctorName)) return [];
  return [
    {
      roleKey: role.key,
      roleLabel: role.label,
      name: text(s.doctorName),
      qualification: text(s.degrees),
      regNo: text(s.regNo),
      external: false,
      partnerId: null,
      fee: role.defaultFee || 0,
    },
  ];
}

/** The team to show: the stored one, or an older case's operative.surgeon as a Surgeon row. */
export function teamOf(k) {
  if (Array.isArray(k?.billing?.team)) return k.billing.team;
  const surgeon = text(k?.operative?.surgeon);
  if (!surgeon) return [];
  const label = state().roles.find((r) => r.key === 'surgeon')?.label || 'Surgeon';
  return [
    { roleKey: 'surgeon', roleLabel: label, name: surgeon, qualification: '', regNo: '', external: false, partnerId: null, fee: 0 },
  ];
}

export const teamFees = (team) => (team || []).reduce((s, m) => s + Math.max(Number(m.fee) || 0, 0), 0);

/* ---------------- admin: outside doctors ---------------- */
function partnerValues(body, row) {
  const st = state();
  const out = {};
  if (body.name != null) {
    const name = text(body.name);
    if (!name) throw httpError(422, 'An outside doctor needs a name');
    if (st.partners.some((p) => p !== row && p.name.toLowerCase() === name.toLowerCase()))
      throw httpError(409, `'${name}' is already on the list of outside doctors`);
    out.name = name;
  }
  ['qualification', 'regNo', 'phone', 'note'].forEach((k) => {
    if (body[k] != null) out[k] = text(body[k]);
  });
  if ('qualification' in out && !out.qualification)
    throw httpError(422, 'An outside doctor needs a medical qualification (e.g. MD Anaesthesia)');
  if ('defaultRoleKey' in body) {
    const key = body.defaultRoleKey || null;
    if (key && !st.roles.some((r) => r.key === key)) throw httpError(422, `Unknown team role '${key}'`);
    out.defaultRoleKey = key;
  }
  if (body.defaultFee != null) out.defaultFee = fee(body.defaultFee, 'Usual fee');
  if (body.active != null) out.active = !!body.active;
  return out;
}

export const otTeam = {
  async options() {
    await latency(20);
    const st = state();
    return c({
      roles: ordered(st.roles).filter((r) => r.active),
      partners: byName(st.partners).filter((p) => p.active),
      staff: S.staff
        .filter((s) => s.active !== false && STAFF_ROLES_IN_OT.includes(s.role))
        .map((s) => ({ name: s.name, role: s.role }))
        .sort((a, b) => a.name.toLowerCase().localeCompare(b.name.toLowerCase())),
    });
  },
  admin: {
    async roles() {
      await latency(20);
      return c(ordered(state().roles));
    },
    async createRole(label, defaultFee = 0) {
      await latency(30);
      const rows = state().roles;
      const n = text(label);
      if (!n) throw httpError(422, 'A team role needs a name');
      if (rows.some((r) => r.label.toLowerCase() === n.toLowerCase())) throw httpError(409, `'${n}' is already on the list`);
      const base =
        n
          .toLowerCase()
          .replace(/[^a-z0-9]+/g, '_')
          .replace(/^_+|_+$/g, '')
          .slice(0, 30) || 'ot_team_role';
      let key = base;
      for (let i = 2; rows.some((r) => r.key === key); i += 1) key = `${base.slice(0, 27)}_${i}`;
      const row = {
        id: Math.max(0, ...rows.map((r) => r.id)) + 1,
        key,
        label: n,
        defaultFee: fee(defaultFee, 'Usual fee'),
        sortOrder: Math.max(0, ...rows.map((r) => r.sortOrder)) + 1,
        active: true,
      };
      rows.push(row);
      store.notify();
      return c(row);
    },
    async updateRole(key, patch = {}) {
      await latency(30);
      const rows = state().roles;
      const row = rows.find((r) => r.key === key);
      if (!row) throw httpError(404, 'Team role not found');
      if (patch.label != null) {
        const n = text(patch.label);
        if (!n) throw httpError(422, 'A team role needs a name');
        if (rows.some((r) => r !== row && r.label.toLowerCase() === n.toLowerCase()))
          throw httpError(409, `'${n}' is already on the list`);
        row.label = n;
      }
      if (patch.defaultFee != null) row.defaultFee = fee(patch.defaultFee, 'Usual fee');
      if (patch.active != null) row.active = !!patch.active;
      store.notify();
      return c(row);
    },
    async reorderRoles(keys) {
      await latency(30);
      const rows = state().roles;
      if (keys.length !== rows.length || !rows.every((r) => keys.includes(r.key)))
        throw httpError(422, 'order must list every existing ot team role exactly once');
      keys.forEach((k, i) => {
        rows.find((r) => r.key === k).sortOrder = i;
      });
      store.notify();
      return c(ordered(rows));
    },
    async partners() {
      await latency(20);
      return c(byName(state().partners));
    },
    async createPartner(body = {}) {
      await latency(30);
      const st = state();
      const clean = partnerValues({ ...body, name: body.name ?? '', qualification: body.qualification ?? '' });
      const row = {
        id: Math.max(0, ...st.partners.map((p) => p.id)) + 1,
        regNo: '',
        phone: '',
        note: '',
        defaultRoleKey: null,
        defaultFee: 0,
        ...clean,
        active: true,
        createdAt: new Date().toISOString(),
      };
      st.partners.push(row);
      store.notify();
      return c(row);
    },
    async updatePartner(id, patch = {}) {
      await latency(30);
      const row = state().partners.find((p) => p.id === Number(id));
      if (!row) throw httpError(404, 'Outside doctor not found');
      Object.assign(row, partnerValues(patch, row));
      store.notify();
      return c(row);
    },
  },
};

File: backend/app/services/ot_team.py
"""OT team: who was in the operating theatre for a surgery, including outside doctors / partners
(visiting surgeon, anaesthetist... not clinic staff) with their medical qualification, and the fee
each one adds to the surgery's bill.

Lists (admin-editable, switched off rather than deleted so old cases keep them):
  OtTeamRole  — Surgeon, Assistant surgeon, Anaesthetist... with the usual fee for that role.
  OtPartner   — the directory of outside doctors: name, qualification, reg. no., phone, usual role
                and usual fee.

On a case the team is `billing.team` (billing is the section every staff role may edit, and the
fees are billing): [{roleKey, roleLabel, name, qualification, regNo, external, partnerId, fee}].
Role label, name, qualification and reg. no. are copies (snapshots) — renaming a role or an outside
doctor later does not change old records. `clean_team` validates and tidies a team on every save.

Every admin write appends an AuditLog row. Errors are app.services.admin NotFound / Conflict /
BadValue (404 / 409 / 422 in the routes).
"""
import re

from sqlalchemy import func, select
from sqlalchemy.orm import Session

from app.models.ot import OtCase, OtPartner, OtTeamRole
from app.models.staff import Staff
from app.seed.ot_team import SURGEON_KEY
from app.services import rx_print
from app.services.admin import BadValue, Conflict, NotFound

STAFF_ROLES_IN_OT = ("doctor", "ot_staff", "optometrist")
MAX_NAME = 120
MAX_QUAL = 120
MAX_REG = 60


def _text(value, what: str, limit: int) -> str:
    s = re.sub(r"\s+", " ", str(value or "")).strip()
    if len(s) > limit:
        raise BadValue(f"{what}: keep it under {limit} characters")
    return s


def _fee(value, where: str) -> int:
    if value is None or value == "":
        return 0
    if isinstance(value, bool):
        raise BadValue(f"{where}: the fee must be a number of rupees")
    if isinstance(value, float):
        if not value.is_integer():
            raise BadValue(f"{where}: the fee must be whole rupees (no paise)")
        value = int(value)
    if isinstance(value, str):
        s = value.strip().replace(",", "").removeprefix("₹").strip()
        if not re.fullmatch(r"-?\d+", s):
            raise BadValue(f"{where}: the fee must be whole rupees, e.g. 2500")
        value = int(s)
    if not isinstance(value, int):
        raise BadValue(f"{where}: the fee must be a number of rupees")
    if value < 0:
        raise BadValue(f"{where}: the fee cannot be below ₹0")
    return value


# --------------------------------------------------------------------------- lists
def roles(db: Session, include_inactive: bool = False) -> list[OtTeamRole]:
    stmt = select(OtTeamRole).order_by(OtTeamRole.sort_order, OtTeamRole.id)
    if not include_inactive:
        stmt = stmt.where(OtTeamRole.active.is_(True))
    return list(db.scalars(stmt))


def partners(db: Session, include_inactive: bool = False) -> list[OtPartner]:
    stmt = select(OtPartner).order_by(func.lower(OtPartner.name), OtPartner.id)
    if not include_inactive:
        stmt = stmt.where(OtPartner.active.is_(True))
    return list(db.scalars(stmt))


def staff_options(db: Session) -> list[dict]:
    rows = db.scalars(select(Staff).where(Staff.active.is_(True), Staff.role.in_(STAFF_ROLES_IN_OT))
                      .order_by(func.lower(Staff.name)))
    return [{"name": s.name, "role": s.role} for s in rows]


def options(db: Session) -> dict:
    return {"roles": roles(db), "partners": partners(db), "staff": staff_options(db)}


# --------------------------------------------------------------------------- admin: roles
def create_role(db: Session, *, label: str, default_fee: int, by) -> OtTeamRole:
    return rx_print.create_row(db, OtTeamRole, label=label, key=None, by=by, what="A team role",
                               entity="ot_team_role", default_fee=_fee(default_fee, "Usual fee"))


def get_role(db: Session, key: str) -> OtTeamRole:
    return rx_print.get_row(db, OtTeamRole, key, "Team role")


def update_role(db: Session, row: OtTeamRole, values: dict, by) -> OtTeamRole:
    changed = {}
    if values.get("label") is not None:
        label = rx_print._clean_label(db, OtTeamRole, values["label"], "A team role", exclude_id=row.id)
        if label != row.label:
            row.label = changed["label"] = label
    if values.get("default_fee") is not None:
        fee = _fee(values["default_fee"], "Usual fee")
        if fee != row.default_fee:
            row.default_fee = changed["default_fee"] = fee
    if values.get("active") is not None and values["active"] != row.active:
        row.active = changed["active"] = values["active"]
    rx_print._audit(db, by, "ot_team_role.update", "ot_team_role", row.id, key=row.key, **changed)
    db.commit()
    return row


def reorder_roles(db: Session, keys: list[str], by) -> list[OtTeamRole]:
    return rx_print.reorder_rows(db, OtTeamRole, keys, by, "ot_team_role")


# --------------------------------------------------------------------------- admin: outside doctors
def get_partner(db: Session, partner_id: int) -> OtPartner:
    row = db.get(OtPartner, partner_id)
    if row is None:
        raise NotFound("Outside doctor not found")
    return row


def _partner_values(db: Session, values: dict, row: OtPartner | None = None) -> dict:
    out = {}
    if values.get("name") is not None:
        name = _text(values["name"], "Name", MAX_NAME)
        if not name:
            raise BadValue("An outside doctor needs a name")
        stmt = select(OtPartner).where(func.lower(OtPartner.name) == name.lower())
        if row is not None:
            stmt = stmt.where(OtPartner.id != row.id)
        if db.scalar(stmt) is not None:
            raise Conflict(f"'{name}' is already on the list of outside doctors")
        out["name"] = name
    for field, what, limit in (("qualification", "Qualification", MAX_QUAL), ("reg_no", "Reg. no.", MAX_REG),
                               ("phone", "Phone", 20), ("note", "Note", 255)):
        if values.get(field) is not None:
            out[field] = _text(values[field], what, limit)
    if "qualification" in out and not out["qualification"]:
        raise BadValue("An outside doctor needs a medical qualification (e.g. MD Anaesthesia)")
    if "default_role_key" in values:
        key = values["default_role_key"] or None
        if key is not None and db.scalar(select(OtTeamRole).filter_by(key=key)) is None:
            raise BadValue(f"Unknown team role '{key}'")
        out["default_role_key"] = key
    if values.get("default_fee") is not None:
        out["default_fee"] = _fee(values["default_fee"], "Usual fee")
    if values.get("active") is not None:
        out["active"] = bool(values["active"])
    return out


def create_partner(db: Session, values: dict, by) -> OtPartner:
    values = {**values, "qualification": values.get("qualification") or ""}
    clean = _partner_values(db, values)
    row = OtPartner(**{"reg_no": "", "phone": "", "note": "", "default_fee": 0, **clean}, active=True)
    db.add(row)
    db.flush()
    rx_print._audit(db, by, "ot_partner.create", "ot_partner", row.id, **clean)
    db.commit()
    return row


def update_partner(db: Session, row: OtPartner, values: dict, by) -> OtPartner:
    clean = _partner_values(db, values, row)
    changed = {}
    for k, v in clean.items():
        if getattr(row, k) != v:
            setattr(row, k, v)
            changed[k] = v
    rx_print._audit(db, by, "ot_partner.update", "ot_partner", row.id, **changed)
    db.commit()
    return row


# --------------------------------------------------------------------------- the team on a case
def _row_empty(r: dict) -> bool:
    return not any(str(r.get(k) or "").strip() for k in ("name", "qualification", "regNo")) and not r.get("fee")


def clean_team(db: Session, team) -> list[dict]:
    """Validate and tidy a case's team list. Fully empty rows are dropped. Each row needs a known
    role and a name; the fee is whole rupees, ₹0 or more; an outside (external) member must have a
    medical qualification. Role label / name / qualification / reg. no. are kept as sent (snapshots);
    a missing role label is filled from the role list. BadValue names the row that is wrong."""
    if team is None:
        return []
    if not isinstance(team, list):
        raise BadValue("The OT team must be a list of team members")
    all_roles = {r.key: r for r in db.scalars(select(OtTeamRole))}
    out = []
    for i, raw in enumerate(team, start=1):
        if not isinstance(raw, dict):
            raise BadValue(f"OT team row {i}: not a team member")
        if _row_empty(raw):
            continue
        key = str(raw.get("roleKey") or "").strip()
        role = all_roles.get(key)
        where = f"OT team row {i}" + (f" ({role.label})" if role else "")
        if not key:
            raise BadValue(f"{where}: pick a role (Surgeon, Anaesthetist…)")
        if role is None:
            raise BadValue(f"{where}: unknown role '{key}'")
        name = _text(raw.get("name"), f"{where} name", MAX_NAME)
        if not name:
            raise BadValue(f"{where}: enter the person's name")
        where = f"OT team row {i} ({role.label}, {name})"
        qualification = _text(raw.get("qualification"), f"{where} qualification", MAX_QUAL)
        external = bool(raw.get("external"))
        if external and not qualification:
            raise BadValue(f"{where}: an outside doctor needs a medical qualification (e.g. MD Anaesthesia)")
        partner_id = raw.get("partnerId")
        if partner_id in ("", None):
            partner_id = None
        else:
            try:
                partner_id = int(partner_id)
            except (TypeError, ValueError):
                raise BadValue(f"{where}: unknown outside doctor")
            if db.get(OtPartner, partner_id) is None:
                raise BadValue(f"{where}: unknown outside doctor")
        out.append({
            "roleKey": key,
            "roleLabel": _text(raw.get("roleLabel"), f"{where} role", 80) or role.label,
            "name": name,
            "qualification": qualification,
            "regNo": _text(raw.get("regNo"), f"{where} reg. no.", MAX_REG),
            "external": external,
            "partnerId": partner_id,
            "fee": _fee(raw.get("fee"), where),
        })
    return out


def default_team(db: Session) -> list[dict]:
    """A new case starts with the clinic's doctor as Surgeon (name, degrees and reg. no. from the
    prescription print settings), at the Surgeon role's usual fee."""
    role = db.scalar(select(OtTeamRole).filter_by(key=SURGEON_KEY))
    settings = rx_print.get_settings(db)
    name = (settings.get("doctorName") or "").strip()
    if role is None or not role.active or not name:
        return []
    return [{"roleKey": role.key, "roleLabel": role.label, "name": name,
             "qualification": (settings.get("degrees") or "").strip(), "regNo": (settings.get("regNo") or "").strip(),
             "external": False, "partnerId": None, "fee": role.default_fee or 0}]


def team_of(db: Session | None, case_or_billing, operative: dict | None = None) -> list[dict]:
    """The team to show: the stored one, or for an older case saved before teams existed, its
    `operative.surgeon` as a Surgeon row (read only; nothing is written)."""
    if isinstance(case_or_billing, OtCase):
        billing, operative = case_or_billing.billing or {}, case_or_billing.operative or {}
    else:
        billing = case_or_billing or {}
    if isinstance(billing.get("team"), list):
        return billing["team"]
    surgeon = ((operative or {}).get("surgeon") or "").strip()
    if not surgeon:
        return []
    label = "Surgeon"
    if db is not None:
        role = db.scalar(select(OtTeamRole).filter_by(key=SURGEON_KEY))
        label = role.label if role else label
    return [{"roleKey": SURGEON_KEY, "roleLabel": label, "name": surgeon, "qualification": "", "regNo": "",
             "external": False, "partnerId": None, "fee": 0}]


def team_fees(team: list[dict] | None) -> int:
    total = 0
    for r in team or []:
        try:
            total += max(int(r.get("fee") or 0), 0)
        except (TypeError, ValueError):
            pass
    return total
